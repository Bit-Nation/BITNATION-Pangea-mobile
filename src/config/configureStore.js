// @flow

import { createStore, applyMiddleware, compose, type Store } from 'redux';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import createSagaMonitor from '@clarketm/saga-monitor';
import reducers from '../reducers';
import rootSaga from '../sagas';

/**
 * @desc Configures a Redux store.
 * @return {Store} Created store object.
 */
export default function configureStore(): Store {
  const sagaMonitor = createSagaMonitor({
    level: 'log',
    actionDispatch: true,
  });
  const sagaMiddleware = createSagaMiddleware({ sagaMonitor });

  const enhancer = compose(
    applyMiddleware(sagaMiddleware),
    applyMiddleware(logger),
  );

  const store = createStore(reducers, enhancer);
  sagaMiddleware.run(rootSaga);
  return store;
}
