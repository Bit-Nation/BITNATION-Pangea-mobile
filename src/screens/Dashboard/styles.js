// @flow

import Colors from '../../global/colors';
import GlobalStyles from '../../global/Styles';

const styles = {
  ...GlobalStyles,

  panelViewTitle: {
    ...GlobalStyles.panelViewTitle,
    color: Colors.BitnationHighlightColor,
  },
  activityPanelContainer: {
    flex: 1,
  },
  activityPanel: {
    flex: 1,
  },
  bottomContainer: {
    flex: 2,
    flexDirection: 'row',
  },
  nationsPanel: {
    flex: 1,
    marginRight: 4,
  },
  rightContainer: {
    marginLeft: 4,
    flex: 1,
    // flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'stretch',
  },
  walletPanel: {
    // flex: 2,
  },
  // This is a panel, unlike the other styles above, which are views
  // which hold panelViews.
  warningPanel: {
    ...GlobalStyles.gridPanelView,
    flex: 1,
  },
  warningPanelBody: {
    ...GlobalStyles.body,
    fontWeight: '300',
  },
  //
  confirmKeyPanel: {
    ...GlobalStyles.gridPanelView,
    flex: 1,
  },
  confirmKeyBody: {
    ...GlobalStyles.body,
    fontWeight: '500',
  },
};
export default styles;
