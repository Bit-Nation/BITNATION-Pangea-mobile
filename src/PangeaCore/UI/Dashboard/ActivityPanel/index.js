// @flow

import React from 'react';
import {
  FlatList,
  Text,
  View,
} from 'react-native';

import styles from './styles';
import PanelView from '../../../components/common/PanelView';
import i18n from '../../../global/i18n';
import Button from '../../../components/common/Button';
import type { ActivityLogMessage } from '../../../types/ActivityLogMessage';

type Props = {
  /**
   * @desc Array of messages objects to display.
   */
  +messages: Array<ActivityLogMessage>,
  /**
   * @desc Style to be applied to root view.
   */
  style: any,
};

type TestingModeProps = {
  /**
   * @desc State of testing mode. For testing only.
   */
  testingMode: {
    +isActive: bool,
  },
  /**
   * @desc Callback to add dummy message. For testing only.
   */
  onAddDummyMessage: () => void,
}

/**
 * @desc Component to render nation panel on dashboard
 * @return {React.Component} A component.
 */
const ActivityPanel = ({
  style, messages, testingMode, onAddDummyMessage,
}: Props & TestingModeProps) => ((
  <View style={style}>
    <PanelView
      style={styles.gridPanelView}
      titleStyle={styles.panelViewTitle}
      childrenContainerStyle={styles.flex}
      title={i18n.t('screens.dashboard.activityPanel.title')}
    >
      <FlatList
        renderItem={(item: { item: ActivityLogMessage }) => {
          const message = item.item;
          const messageText = message.interpret ? i18n.t(`activityLog.${message.msg}`, JSON.parse(message.params)) : message.msg;
          return (<Text style={styles.listItemText}>{messageText}</Text>);
        }}
        keyExtractor={(item: ActivityLogMessage) => item.id.toString()}
        data={messages}
        style={styles.flex}
      />
      {
        testingMode.isActive === true &&
        <Button title='Add dummy log' onPress={onAddDummyMessage} />
      }

    </PanelView>
  </View>
));

ActivityPanel.defaultProps = {
  messages: [],
};

export default ActivityPanel;
