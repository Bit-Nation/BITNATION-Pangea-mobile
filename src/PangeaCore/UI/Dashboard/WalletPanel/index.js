// @flow

import React from 'react';
import { View, Text } from 'react-native';
import _ from 'lodash';

import styles from './styles';

import PanelView from '../../../components/common/PanelView';
import i18n from '../../../global/i18n';
import { prettyWalletBalance } from '../../../utils/formatters';
import { resolveWallet } from '../../../utils/wallet';
import type { WalletType } from '../../../types/Wallet';

type Props = {
  /**
   * @desc Array of wallet objects.
   */
  +wallets: Array<WalletType>,
  /**
   * @desc Style to be applied to root view.
   */
  style: any,
};

/**
 * @desc Component to render wallet panel on dashboard
 * @return {React.Component} A component.
 */
const WalletPanel = ({ style, wallets } : Props) => {
  const patWallet = resolveWallet(wallets, 'XPAT');
  if (patWallet === null) {
    return <View />;
  }
  return (
    <View style={style}>
      <PanelView
        style={styles.walletGridPanel}
        childrenContainerStyle={styles.noflex}
        title={i18n.t('screens.dashboard.walletPanel.title')}
        titleStyle={styles.panelViewTitle}
      >
        {
          _.isEmpty(wallets)
            ?
              <Text style={styles.body}>{i18n.t('screens.dashboard.walletPanel.empty')}</Text>
            :
              <View>
                <Text style={styles.footnote}>{i18n.t('common.bitnationPat')}</Text>
                <Text style={[styles.largeTitle, styles.textWallet]}>{prettyWalletBalance(patWallet)}</Text>
              </View>
        }
      </PanelView>
    </View>
  );
};

WalletPanel.defaultProps = {
  wallets: [],
};

export default WalletPanel;
