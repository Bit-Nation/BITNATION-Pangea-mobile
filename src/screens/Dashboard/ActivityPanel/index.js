import React, { Component } from 'react';
import {
  FlatList,
  Text,
  View,
} from 'react-native';
import PropTypes from 'prop-types';

import styles from './styles';
import PanelView from '../../../components/common/PanelView';
import i18n from '../../../global/i18n';
import Button from '../../../components/common/Button';

/**
 * @desc Component to render nation panel on dashboard
 * @type React.Component
 */
export default class ActivityPanel extends Component {

  render() {
    const { style } = this.props;

    return (
      <View style={style}>
        <PanelView style={styles.gridPanelView}
                   titleStyle={styles.panelViewTitle}
                   childrenContainerStyle={styles.flex}
                   title={i18n.t('screens.dashboard.activityPanel.title')}>
          <FlatList
            renderItem={(item) => {
              const message = item.item;
              const messageText = message.interpret ? i18n.t(`activityLog.${message.msg}`, JSON.parse(message.params)) : message.msg;
              return (<Text style={styles.listItemText}>{messageText}</Text>);
            }}
            keyExtractor={(item) => item.id}
            data={this.props.messages}
            style={styles.flex}
          />
          {
            this.props.testingMode.isActive &&
            <Button title='Add dummy log' onPress={this.props.onAddDummyMessage}/>
          }

        </PanelView>
      </View>
    );
  }

}

ActivityPanel.propTypes = {
  /**
   * @desc Array of messages objects to display
   */
  messages: PropTypes.array,
  /**
   * @desc State of testing mode. For testing only.
   */
  testingMode: PropTypes.object,
  /**
   * @desc Callback to add dummy message. For testing only.
   */
  onAddDummyMessage: PropTypes.func,
};

ActivityPanel.defaultProps = {
  messages: [],
};
