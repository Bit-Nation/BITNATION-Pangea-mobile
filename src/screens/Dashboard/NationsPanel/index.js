// @flow

import React from 'react';
import {
  FlatList,
  Text,
  View,
} from 'react-native';
import _ from 'lodash';

import styles from './styles';
import PanelView from '../../../components/common/PanelView';
import i18n from '../../../global/i18n';
import NationListItem from '../../../components/common/NationListItem';
import { nationIsDraft } from '../../../utils/nations';
import Loading from '../../../components/common/Loading';
import type { NationIdType, NationType } from '../../../types/Nation';

const NEWEST_NATION_COUNT = 5;

type Props = {
  /**
   * @desc Array of all nations objects
   */
  +nations: Array<NationType>,
  /**
   * @desc Callback on select nation to open. Takes one parameter - id of nation.
   */
  onSelectNation: (NationIdType) => void,
  /**
   * @desc Flag if loading is in progress.
   */
  loadingInProgress: boolean,
  /**
   * @desc Style to be applied to root view.
   */
  style: any,
};

/**
 * @desc Component to render nation panel on dashboard
 * @return {React.Component} A component
 */
const NationsPanel = ({
  style, nations, onSelectNation, loadingInProgress,
}: Props) => {
  const existedNations = _.filter(nations, nation => nationIsDraft(nation) === false);
  const newestNations = _.take(_.sortBy(existedNations, nation => -nation.idInSmartContract), NEWEST_NATION_COUNT);
  const nationsCountStrings = i18n.t('screens.dashboard.nationsPanel.nationsCount', { count: existedNations.length });

  return (
    <View style={style}>
      <PanelView
        style={styles.nationsGridPanel}
        titleStyle={styles.panelViewTitle}
        title={i18n.t('screens.dashboard.nationsPanel.title')}
      >
        <View style={styles.nationsCountContainer}>
          <Text style={styles.body}>{nationsCountStrings.prefix}
            <Text style={styles.nationsCountString}>{nationsCountStrings.main}</Text>
            {nationsCountStrings.suffix}
          </Text>
        </View>

        {/* Nations list. This view extends the margins to edges */}
        <View style={styles.listContainer}>

          <View style={styles.panelFlatlistHeader}>
            <Text
              style={styles.nationsListHeaderText}
            >{i18n.t('screens.dashboard.nationsPanel.newNations')}
            </Text>
          </View>

          <FlatList
            renderItem={(item) => {
              const nation = item.item;
              return (
                <NationListItem
                  text={nation.nationName}
                  textStyle={styles.nationsListText}
                  onPress={onSelectNation}
                  id={nation.id}
                />);
            }}
            ItemSeparatorComponent={() => <View style={styles.sectionListSeparator} />}
            keyExtractor={item => item.id.toString()}
            data={newestNations}
          />
        </View>
      </PanelView>
      {loadingInProgress && _.isEmpty(nations) && <Loading />}
    </View>
  );
};

NationsPanel.defaultProps = {
  nations: [],
  onSelectNation: () => null,
  loadingInProgress: false,
};

export default NationsPanel;
