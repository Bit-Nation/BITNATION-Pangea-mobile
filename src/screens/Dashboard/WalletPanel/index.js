// @flow

import React from 'react';
import { View, Text } from 'react-native';
import _ from 'lodash';

import styles from './styles';

import PanelView from '../../../components/common/PanelView';
import i18n from '../../../global/i18n';
import { prettyWalletBalance } from '../../../utils/formatters';
import type { WalletType } from '../../../types/Wallet';

type Props = {
  /**
   * @desc Array of wallet objects.
   */
  +wallets: Array<WalletType>,
  /**
   * @desc Style to be applied to root view.
   */
  style: any,
};

/**
 * @desc Component to render wallet panel on dashboard
 * @return {React.Component} A component.
 */
const WalletPanel = ({ style, wallets } : Props) => ((
  <View style={style}>
    <PanelView
      style={styles.walletGridPanel}
      childrenContainerStyle={styles.noflex}
      title={i18n.t('screens.dashboard.walletPanel.title')}
      titleStyle={styles.panelViewTitle}
    >
      {
        _.isEmpty(wallets)
          ?
            <Text style={styles.body}>{i18n.t('screens.dashboard.walletPanel.empty')}</Text>
          :
            <View>
              <Text style={styles.footnote}>XPAT</Text>
              <Text style={styles.currencyLarge}>{prettyWalletBalance(wallets[1])}</Text>
            </View>
      }
    </PanelView>
  </View>
));

WalletPanel.defaultProps = {
  wallets: [],
};

export default WalletPanel;
