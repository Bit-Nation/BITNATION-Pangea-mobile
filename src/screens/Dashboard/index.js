// @flow

import React, { Component } from 'react';
import { View, Text } from 'react-native';
import { connect } from 'react-redux';

import styles from './styles';
import BackgroundImage from '../../components/common/BackgroundImage';
import FakeNavigationBar from '../../components/common/FakeNavigationBar';
import PanelView from '../../components/common/PanelView';
import i18n from '../../global/i18n';
import WalletPanel from './WalletPanel';
import ActivityPanel from './ActivityPanel';
import NationsPanel from './NationsPanel';
import { openNation } from '../../actions/nations';
import { screen } from '../../global/Screens';
import { addNewMessage } from '../../actions/activity';
import type { NationIdType } from '../../types/Nation';
import type { State } from '../../reducers';
import type { Navigator } from '../../types/ReactNativeNavigation';
import { getCurrentAccount } from '../../reducers/accounts';

type Props = {
  /**
   * @desc React Native Navigation navigator object.
   */
  navigator: Navigator,
};

type Actions = {
  /**
   * @desc Callback on nation select.
   * @param {NationIdType} id Id of selected nation.
   */
  onSelectNation: (id: NationIdType) => void
};

type TestingModeProps = {
  /**
   * @desc Function to add dummy log activity message for testing.
   */
  onAddDummyMessage: () => void
};

class Dashboard extends Component<Props & Actions & State & TestingModeProps> {
  onSelectNation = (id) => {
    this.props.onSelectNation(id);
    this.props.navigator.push(screen('NATION_DETAILS_SCREEN'));
  };

  onStartKeyConfirmation = () => {
    this.props.navigator.showModal({
      ...screen('CONFIRM_KEY_INSTRUCTION_SCREEN'),
      passProps: {
        shouldShowCancel: true,
      },
    });
  };

  render() {
    const currentAccount = getCurrentAccount(this.props.accounts);

    return (
      <View
        testID='dashboardView'
        style={styles.screenContainer}
      >
        <BackgroundImage />
        <FakeNavigationBar navBarHidden />
        <View style={styles.gridContainer}>
          <View style={styles.activityPanelContainer}>
            <ActivityPanel
              style={styles.activityPanel}
              messages={this.props.activity.messages}
              testingMode={this.props.testingMode}
              onAddDummyMessage={this.props.onAddDummyMessage}
            />
          </View>
          <View style={styles.bottomContainer}>
            <NationsPanel
              nations={this.props.nations.nations}
              onSelectNation={this.onSelectNation}
              style={styles.nationsPanel}
              loadingInProgress={this.props.nations.inProgress}
            />
            <View style={styles.rightContainer}>
              <WalletPanel
                wallets={this.props.wallet.wallets || []}
                style={styles.walletPanel}
              />
              {
                (currentAccount === null || currentAccount.confirmedMnemonic === true) ?
                  <PanelView
                    title={i18n.t('screens.dashboard.warningPanel.title')}
                    style={styles.warningPanel}
                    titleStyle={styles.panelViewTitle}
                  >
                    <Text style={styles.warningPanelBody}>{i18n.t('screens.dashboard.warningPanel.body')}</Text>
                  </PanelView>
                  :
                  <PanelView
                    title={i18n.t('screens.dashboard.confirmKeyPanel.title')}
                    style={styles.confirmKeyPanel}
                    titleStyle={styles.alertPanelViewTitle}
                    buttonTitle={i18n.t('screens.dashboard.confirmKeyPanel.button')}
                    onButtonClick={this.onStartKeyConfirmation}
                  >
                    <Text style={styles.confirmKeyBody}>{i18n.t('screens.dashboard.confirmKeyPanel.body')}</Text>
                  </PanelView>
              }
            </View>
          </View>
        </View>
      </View >
    );
  }
}

const mapStateToProps = state => ({
  ...state,
});

const mapDispatchToProps = dispatch => ({
  onSelectNation(id) {
    dispatch(openNation(id));
  },
  onAddDummyMessage() {
    dispatch(addNewMessage('dummy message'));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
