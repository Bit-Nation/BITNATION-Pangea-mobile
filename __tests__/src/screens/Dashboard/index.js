import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import Dashboard from '../../../../src/screens/Dashboard';

test('Dashboard renders correctly', () => {
  const initialStateMock = {
    nations: {
      nations: [{
        id: 1,
        idInSmartContract: 1,
      }],
    },
    activity: {
      messages: [],
    },
    testingMode: {
      isActive: false,
    },
    wallet: {
      wallets: [],
    },
  };
  const storeMock = configureStore([]);
  const propsMock = {
    navigator: {
      push: jest.fn(),
    },
  };

  const wrapper = shallow((
    <Dashboard {...propsMock} store={storeMock(initialStateMock)} />
  ));
  expect(wrapper).toMatchSnapshot();
});
