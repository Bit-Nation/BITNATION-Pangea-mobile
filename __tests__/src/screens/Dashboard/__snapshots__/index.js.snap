// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Dashboard)
    navigator={
      Object {
        "pop": [MockFunction],
        "push": [MockFunction],
        "setButtons": [MockFunction],
        "setOnNavigatorEvent": [MockFunction],
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activity": Object {
        "messages": Array [],
      },
      "nations": Object {
        "nations": Array [
          Object {
            "id": 1,
            "idInSmartContract": 1,
          },
        ],
      },
      "navigator": Object {
        "pop": [MockFunction],
        "push": [MockFunction],
        "setButtons": [MockFunction],
        "setOnNavigatorEvent": [MockFunction],
      },
      "onAddDummyMessage": [Function],
      "onSelectNation": [Function],
      "openDApp": [Function],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
      "testingMode": Object {
        "isActive": false,
      },
      "wallet": Object {
        "wallets": Array [],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activity": Object {
          "messages": Array [],
        },
        "nations": Object {
          "nations": Array [
            Object {
              "id": 1,
              "idInSmartContract": 1,
            },
          ],
        },
        "navigator": Object {
          "pop": [MockFunction],
          "push": [MockFunction],
          "setButtons": [MockFunction],
          "setOnNavigatorEvent": [MockFunction],
        },
        "onAddDummyMessage": [Function],
        "onSelectNation": [Function],
        "openDApp": [Function],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "testingMode": Object {
          "isActive": false,
        },
        "wallet": Object {
          "wallets": Array [],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
